extension/src/test/suite/experiments/index.test.ts

```diff
@@ -1422,6 +1434,73 @@ suite('Experiments Test Suite', () => {
       expect(mockUpdateExperimentsData).to.be.calledOnce
       expect(experimentsModel.getIsBranchesView()).to.be.false
     }).timeout(WEBVIEW_TEST_TIMEOUT)
+
+    it('should handle a message to select branches', async () => {
+      const {
+        experiments,
+        experimentsModel,
+        messageSpy,
+        mockUpdateExperimentsData,
+        mockSelectBranches
+      } = setupExperimentsAndMockCommands()
+      const selectBranchesToShowSpy = spy(
+        experimentsModel,
+        'selectBranchesToShow'
+      )
+
+      const webview = await experiments.showWebview()
+      messageSpy.resetHistory()
+      const mockMessageReceived = getMessageReceivedEmitter(webview)
+
+      mockMessageReceived.fire({
+        type: MessageFromWebviewType.SELECT_BRANCHES
+      })
+
+      expect(mockSelectBranches).to.be.calledOnce
+
+      // Needed for the test to pass, but shouldn't be needed.
```

Comment:
Racked my brain a whole day for that. Still can't figure out why it's needed. I really wish I could remove it.

Reply:
This is because waiting for 

```
    const selectedBranches = await this.selectBranches(
      this.experiments.getBranchesToShow()
    )
```

pushes the result into the next event loop, meaning that the assertion executes before the stub is called.

`await Time.delay(0)` is equivalent to `await new Promise(process.nextTick)` so by calling that you delay the assertion until the code under test has completed (FIFO).

I would rewrite the test like this to make it deterministic:

```
    it.only('should handle a message to select branches', async () => {
      const {
        experiments,
        experimentsModel,
        messageSpy,
        mockUpdateExperimentsData,
        mockSelectBranches
      } = setupExperimentsAndMockCommands()
      const mockSelectBranchesToShow = stub(
        experimentsModel,
        'selectBranchesToShow'
      )

      const waitForBranchesToBeSelected = new Promise(resolve =>
        mockSelectBranchesToShow.callsFake(() => resolve(undefined))
      )

      const webview = await experiments.showWebview()
      messageSpy.resetHistory()
      const mockMessageReceived = getMessageReceivedEmitter(webview)

      mockMessageReceived.fire({
        type: MessageFromWebviewType.SELECT_BRANCHES
      })

      expect(mockSelectBranches).to.be.calledOnce

      await waitForBranchesToBeSelected

      expect(mockSelectBranchesToShow).to.be.calledOnceWith(['main', 'other'])

      expect(mockUpdateExperimentsData).to.be.calledOnce
    }).timeout(WEBVIEW_TEST_TIMEOUT)
```

Hope that helps... I used to get stuck on this all the time.

Reply:
You'll see example of this all over the integration tests
```
      const waitForSomeFunctionCall = new Promise(resolve =>
        mockSomeFunction.callsFake(() => resolve(undefined))
      )
      ...
      await waitForSomeFunctionCall
      expect(....
```

Reply:
Thank you so much